data/mods/pokebilities/abilities.js

```diff
@@ -2,38 +2,88 @@
+			let bannedAbilities = ['battlebond', 'comatose', 'disguise', 'flowergift', 'forecast', 'illusion', 'imposter', 'multitype', 'powerconstruct', 'powerofalchemy', 'receiver', 'rkssystem', 'schooling', 'shieldsdown', 'stancechange', 'trace', 'wonderguard', 'zenmode'];
+			bannedAbilities.push(pokemon.ability);
+			if (pokemon.innates) bannedAbilities = bannedAbilities.concat(pokemon.innates);
+			possibleAbilities = possibleAbilities.filter(val => !bannedAbilities.includes(val));
+			if (!possibleAbilities.length) return;
+			let ability = this.getAbility(possibleAbilities[this.random(possibleAbilities.length)]);
+			this.add('-ability', pokemon, ability, '[from] ability: Receiver', '[of] ' + ally);
+			if (isAbility) {
+				pokemon.setAbility(ability);
+			} else {
+				pokemon.removeVolatile("abilityreceiver");
+				pokemon.addVolatile("ability" + ability, pokemon);
+			}
+		},
+	},
 	trace: {
 		inherit: true,
 		onUpdate(pokemon) {
 			if (!pokemon.isStarted) return;
-			let isAbility = this.effect.effectType === "Ability";
-			/**@type {string[]} */
-			let possibleInnates = [];
+			let isAbility = pokemon.ability === 'trace';
 			/**@type {Pokemon[]} */
 			let possibleTargets = [];
 			for (let target of pokemon.side.foe.active) {
-				if (target && !target.fainted && target.innates) {
-					possibleInnates = possibleInnates.concat(target.innates);
-					possibleTargets = possibleTargets.concat(target.innates.map(innate => target));
+				if (target && !target.fainted) {
+					possibleTargets.push(target);
 				}
 			}
-			while (possibleInnates.length) {
-				let rand = 0;
-				if (possibleInnates.length > 1) rand = this.random(possibleInnates.length);
-				let ability = possibleTargets[rand].ability;
-				let innate = possibleInnates[rand];
+			while (possibleTargets.length) {
+				let rand = this.random(possibleTargets.length);
+				let target = possibleTargets[rand];
+				/**@type {string[]} */
+				let possibleAbilities = [target.ability];
+				if (target.innates) possibleAbilities = possibleAbilities.concat(target.innates);
 				let bannedAbilities = ['battlebond', 'comatose', 'disguise', 'flowergift', 'forecast', 'illusion', 'imposter', 'multitype', 'powerconstruct', 'powerofalchemy', 'receiver', 'rkssystem', 'schooling', 'shieldsdown', 'stancechange', 'trace', 'zenmode'];
-				if (bannedAbilities.includes(isAbility ? ability : innate)) {
-					possibleInnates.splice(rand, 1);
+				bannedAbilities.push(pokemon.ability);
+				if (pokemon.innates) bannedAbilities = bannedAbilities.concat(pokemon.innates);
+				possibleAbilities = possibleAbilities.filter(val => !bannedAbilities.includes(val));
+				if (!possibleAbilities.length) {
 					possibleTargets.splice(rand, 1);
 					continue;
 				}
-				this.add('-ability', pokemon, isAbility ? ability : innate, '[from] ability: Trace', '[of] ' + possibleTargets[rand]);
+				rand = this.random(possibleAbilities.length);
```

Comment:
This doesn't work because you still need `rand` to know which target you just Traced. (You can probably call `this.sample` and skip the intermediate variable anyway, here and also for the other Abilities.)